viewing_party/party.py

```diff
@@ -1,23 +1,219 @@
+    
+    user_not_watched = []
+    for friends_movies in clean_all_friends_movies:
+        if friends_movies not in user_data["watched"]:
+            user_not_watched.append(friends_movies)
+    return user_not_watched
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+
+    all_friends_movies = []
+    i = 0
+    for friend in user_data["friends"]:
+        for friends_movie in user_data["friends"][i]["watched"]:
+            all_friends_movies.append(friends_movie)
+        i += 1
+
+    friends_and_user = []
+    for friend_movies in all_friends_movies:
+        if friend_movies not in user_data["watched"]:
+            friends_and_user.append(friend_movies)
+
+    recomended_movies = []
+    j = 0
+    for movie in friends_and_user:
+        for subscription in user_data["subscriptions"]:
+            if subscription in friends_and_user[j]["host"]:
+                recomended_movies.append(movie)
+        j += 1
+
+    return recomended_movies
+
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+from collections import Counter
+
+def get_new_rec_by_genre(user_data):
+
+    if len(user_data["watched"]) == 0:
+        return []
+
+    user_genre = []   
+    for user_movies in user_data["watched"]:
+        user_genre.append(user_movies["genre"])
+        
+        genre_count = Counter(user_genre)
+        user_freq_genre = genre_count.most_common(1)[0][0]
+
+    all_friends_movies = []
+    i = 0
+    for friend in user_data["friends"]:
+        for friend_movie in user_data["friends"][i]["watched"]:
+            all_friends_movies.append(friend_movie)
+        i += 1
+
+    user_not_watched = []
+    for friends_movies in all_friends_movies:
+        if friends_movies not in user_data["watched"]:
+            user_not_watched.append(friends_movies)
+
+    recommended_movies = []
+    i = 0
+    for movie in user_not_watched:
+        if user_freq_genre in user_not_watched[i]["genre"]:
+            recommended_movies.append(movie)
+        i += 1
+
+    return recommended_movies
+
+def get_rec_from_favorites(user_data):
+
+    i = 0
+    for friend in user_data["friends"]:
+        if len(user_data["friends"][i]["watched"]) == 0:
+            return []
+        i += 1
+
+    all_friends_movies = []
+    i = 0
+    for friend in user_data["friends"]:
+        for friend_movie in user_data["friends"][i]["watched"]:
+            all_friends_movies.append(friend_movie)
+        i += 1
+
+    friend_titles = []
+    for friend_movie in all_friends_movies:
+        friend_titles.append(friend_movie["title"])
 
+    friends_not_watched = []
+    for user_movie in user_data["favorites"]:
+        if user_movie["title"] not in friend_titles:
+            friends_not_watched.append(user_movie)
+    
+    return friends_not_watched

```

Comment:
For Waves 4 and 5, there are more straightforward ways of implementing those functions. `get_available_recs` should be able to leverage `get_friends_unique_watched`. With that, all you have to do is iterate over the resulting list and check whether the current item's (movie) host is in the user's subscriptions. If so, add it to a list.

In Wave 5, `get_new_rec_by_genre` can again leverage `get_friends_unique_watched` as well as `get_most_watched_genre`. From there you only need a single for loop with one conditional to create the recs. Lastly, `get_rec_from_favorites`, can leverage `get_unique_watched` and from there it's also a single for-loop with a conditional.